fix($compile): sanitize special chars in directive name

curlydevil opened the pull request:
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
Fix for 


**What is the current behavior? (You can also link to an open issue here)**
In HTML markup only directive name's `x-` and `data-` prefixes are being sanitized. (#16278)


**What is the new behavior (if this is a feature change)?**
Also special chars are being sanitized, e.g. `<div _t></div>` will match `t` directive once again


**Does this PR introduce a breaking change?**
Should not


**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**Other information**:
@petebacondarwin suggested to update the regex, and I hope that I did it in a way he was thinking about it.

Narretz commented:
There some test failures with jquery: https://travis-ci.org/angular/angular.js/jobs/295785053

curlydevil commented:
@Narretz yes, I'm trying to understand why I have them passing locally at 4 browsers but not in CI...

Narretz commented:
Are you sure you are running `tests:jquery`?

curlydevil commented:
I was running `grunt autotest` to check in several browsers, and whole `grunt test` before push...

petebacondarwin commented:
I wonder if it is trying to convert `_t` to `T` when it should be `t` or vice versa?

curlydevil commented:
Checked how `<_t></_t>` behaves with 1.5.x - looks like it did not work this way, so I've dropped corresponding test, just leaving it as a `A` restriction check.

curlydevil commented:
Doh... `Error: Error updating Sauce pass/fail status: 'Could not send request: connect ETIMEDOUT 162.222.75.227:443'`
Is there a possibility to restart single stage or the branch build?

curlydevil committed:
fix($compile): special chars are sanitized from directive name in markup

this fixes bug introduced in 1.6.x
when directive name with preceeding special char
in HTML markup does not match the registered name

Narretz commented:
> Checked how <_t></_t> behaves with 1.5.x - looks like it did not work this way, so I've dropped corresponding test, just leaving it as an A restriction check.

Aren't we creating an inconsistent and undocumented behavior? Because afaik there's nothing about the handling of special characters at the start of a directive name in the docs.
Imo this looks like it was never explicitly supported, and if we support it it should be consistent and documented.

The docs:

>>> The normalization process is as follows:
>>> Strip x- and data- from the front of the element/attributes.
>>> Convert the :, -, or _-delimited name to camelCase.

It doesn't say that :, -, or _ are stripped from the front.

gkalpak commented on src/ng/compile.js:
```diff
@@ -3637,7 +3637,7 @@ function SimpleChange(previous, current) {
 
 
-var PREFIX_REGEXP = /^((?:x|data)[:\-_])/i;
+var PREFIX_REGEXP = /^((?:x|data)*[:\-_])/i;
```
Also, `PREFIX_REGEXP` is used to process `ng-attr-` attributes ([here][1] and this change affects that too.
IMO, this PR should restore the behavior of `directiveNormalize()` pre-73050cdda and not affect anything else.

[1]: https://github.com/curlydevil/angular.js/blob/6a20b352dba085c8e9f03c0c44ab67a4dd27e244/src/ng/compile.js#L2197

curlydevil commented:
@gkalpak looking through usage of new function, its implementation is actually correct for CSS, so changing it back might break the desired behaviour there... should I add yet another camelcasing method?

curlydevil committed:
fix($compile): revert directiveNormalize behaviour to pre-1.6.x

gkalpak commented on src/ng/compile.js:
```diff
@@ -3638,16 +3638,28 @@ SimpleChange.prototype.isFirstChange = function() { return this.previousValue ==
+    .replace(SPECIAL_CHARS_REGEXP, function(_, separator, letter, offset) {
+      return offset ? letter.toUpperCase() : letter;
+    })
+    .replace(MOZ_HACK_REGEXP, 'Moz$1');
```
AFAICT, this was only necessary for CSS, so we don't need it here any more.

gkalpak commented on src/ng/compile.js:
```diff
@@ -3638,16 +3638,28 @@ SimpleChange.prototype.isFirstChange = function() { return this.previousValue ==
+ * Also there is special case for Moz prefix starting with upper case letter.
+ * @param name Name to normalize
+ */
+function camelCase(name) {
```
Since this function is only used in `directiveNormalize()` it makes sense to inline the code there.

gkalpak commented:
> should I add yet another camelcasing method?

More or less. I.e. you just need to update `directiveNormalize()` to have the desired behavior (i.e. the bahavior it had for normalizing directive names before 73050cd.

curlydevil committed:
fix($compile): code review remarks fixed

curlydevil commented:
@gkalpak if current implementation is acceptable, please, approve it or un-require changes =)

gkalpak commented on src/ng/compile.js:
```diff
@@ -3638,7 +3638,7 @@ SimpleChange.prototype.isFirstChange = function() { return this.previousValue ==
 
 var PREFIX_REGEXP = /^((?:x|data)[:\-_])/i;
-var SPECIAL_CHARS_REGEXP = /[:\-_]+(.)/g;
+var SPECIAL_CHARS_REGEXP = /([:\-_]+(.))/g;
```
Why the extra parenthesis? They are not needed afaict (and slow the RegExp down - theoretically :grinning:)

gkalpak commented on src/ng/compile.js:
```diff
@@ -3647,7 +3647,9 @@ var SPECIAL_CHARS_REGEXP = /[:\-_]+(.)/g;
   return name
     .replace(PREFIX_REGEXP, '')
-    .replace(SPECIAL_CHARS_REGEXP, fnCamelCaseReplace);
+    .replace(SPECIAL_CHARS_REGEXP, function(_, separator, letter, offset) {
```
Adjust arguments after removing extra parenthesis from RegExp (see previous comment).

curlydevil committed:
fix($compile): regex fix

Narretz commented on test/ng/compileSpec.js:
```diff
@@ -294,6 +294,26 @@ describe('$compile', function() {
       inject(function($compile) {});
     });
 
+    it('should omit special chars before processing directive name', function() {
```
If this applies only to directives with restrict: 'A', then the test description should reflect that. Maybe even link to this issue in a comment.

curlydevil committed:
fix($compile): added comment to test with link to issue

curlydevil commented:
@Narretz @gkalpak @petebacondarwin any other ideas for improvement?

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -294,6 +294,27 @@ describe('$compile', function() {
+        });
+      });
+      inject(function($compile, $rootScope, log) {
+        element = $compile('<div _t></div>')($rootScope);
```
The `element = ` parts are redundant.

gkalpak commented on test/ng/compileSpec.js:
```diff
@@ -294,6 +294,27 @@ describe('$compile', function() {
       inject(function($compile) {});
     });
 
+    it('should omit special chars before processing attribute directive name', function() {
```
I would change "omit" to "ignore", because I think it bettr describes the behavior.

curlydevil committed:
fix($compile): review remarks
